src/base/similartype.jl

```diff
@@ -1,93 +1,72 @@
-"""
-`set_indstype` should be overloaded for
-types with structured dimensions,
-like `OffsetArrays` or named indices
-(such as ITensors).
-"""
-function set_indstype(arraytype::Type{<:AbstractArray}, dims::Tuple)
-  return set_ndims(arraytype, NDims(length(dims)))
+function check_similartype_output(type::Type)
+  isconcretetype(type) || throw(
+    ArgumentError(
+      "`similartype` output `$type`, which is not concrete. The corresponding call to `similar` may not be defined or may not be type stable.",
+    ),
+  )
+  return type
 end
 
-function similartype(arraytype::Type{<:AbstractArray}, eltype::Type, ndims::NDims)
-  return similartype(similartype(arraytype, eltype), ndims)
-end
+"""
```

Comment:
```suggestion
@doc """
```

Does this work without? I'm a bit confused since I would guess that this gives `UndefVarError` since `similartype` is defined below, but I might actually be wrong about this.

Reply:
I actually didn't think very hard about that, but the docs say using `f` is equivalent to using `function f end`: https://docs.julialang.org/en/v1.11/manual/documentation/#Functions-and-Methods-2. What does `@doc` do?

Reply:
I think the point is that:
```julia
"..."
f
```
is actually equivalent to:
```julia
"..."
function f end
```
according to https://docs.julialang.org/en/v1.11/manual/documentation/#Functions-and-Methods-2, so it implicitly defines the function. I didn't even really think about it, it is a bit surprising to be honest. But anyway, it definitely works since tests pass and I can see the docstring locally, so I think it is just a matter of style in this case.